hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java

```diff
@@ -0,0 +1,145 @@
+
+import static java.util.Collections.emptyList;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * A helper class to handle the construction of
+ * {@link org.apache.hadoop.ozone.om.helpers.ServiceInfoEx} objects,
+ * and refresh of the things that are necessary for constructing it.
+ */
+public final class ServiceInfoProvider {
+
+  private static final Logger LOG = getLogger(ServiceInfoProvider.class);
+
+  private final OzoneManagerProtocol om;
+  private final CertificateClient certClient;
+
+  private String caCertPem;
+  private List<String> caCertPemList;
+
+  public ServiceInfoProvider(
+      OzoneConfiguration config,
+      OzoneManagerProtocol om,
+      CertificateClient certClient
+  ) throws IOException {
+    this(config, om, certClient, false);
+  }
+
+  public ServiceInfoProvider(
+      OzoneConfiguration config,
+      OzoneManagerProtocol om,
+      CertificateClient certClient,
+      boolean skipInitializationForTesting
+  ) throws IOException {
+    this.om = om;
+    if (new SecurityConfig(config).isSecurityEnabled()
+        && !skipInitializationForTesting) {
+      this.certClient = certClient;
+      Set<X509Certificate> certs = getCACertificates();
+      caCertPem = toPEMEncodedString(newestOf(certs));
+      caCertPemList = toPEMEncodedStrings(certs);
+      this.certClient.registerRootCARotationListener(onRootCAChange());
+    } else {
+      this.certClient = null;
+      caCertPem = null;
+      caCertPemList = emptyList();
+    }
+  }
+
+  private Function<List<X509Certificate>, CompletableFuture<Void>>
+      onRootCAChange() {
+    return certs -> {
+      CompletableFuture<Void> returnedFuture = new CompletableFuture<>();
+      try {
+        caCertPem = toPEMEncodedString(newestOf(certs));
+        caCertPemList = toPEMEncodedStrings(certs);
+        returnedFuture.complete(null);
+      } catch (Exception e) {
+        LOG.info("Unable to refresh cached PEM formatted CA " +
+            "certificates.", e);
+        returnedFuture.completeExceptionally(e);
+      }
+      return returnedFuture;
+    };
+  }
+
+  public ServiceInfoEx provide() throws IOException {
+    return new ServiceInfoEx(om.getServiceList(), caCertPem, caCertPemList);
+  }
+
+  private Set<X509Certificate> getCACertificates() {
+    return !certClient.getAllRootCaCerts().isEmpty()
+        ? certClient.getAllRootCaCerts()
+        : certClient.getAllCaCerts();
+  }
+
+  private X509Certificate newestOf(Collection<X509Certificate> certs) {
+    return certs.stream()
+        .max((c1, c2) -> expiry(c1).compareTo(expiry(c2)))
+        .orElse(null);
+  }
+
+  private String toPEMEncodedString(X509Certificate cert)
+      throws SCMSecurityException {
+    return cert == null ? null : CertificateCodec.getPEMEncodedString(cert);
+  }
+
+  private String toPEMEncodedStringUnsafe(X509Certificate cert) {
+    try {
+      return toPEMEncodedString(cert);
+    } catch (SCMSecurityException e) {
+      return null;
```

Comment:
@fapifta , I think we should not hide the exception.

Reply:
Hiding this is not really a problem. The following logic is applied on the client side: if there are a list of CA certs, then the list is being used, if there is no element in the list or the list is null, then the sole caCertPEM is being used. This logic is there to ensure compatibility between the new (1.2+) client and an older server side, so we need to keep that semantic.

If any/all certificates in the list fails to translate to PEM format, then we have a second chance.

I have added an error level log message to sign that there is a problem, but chances are that this will never happen, as we get the in memory X509Certificate already from a PEM formatted string, so we should fail earlier if there is a problem with this conversion.

Reply:
@fapifta , thanks for the explanation. 
It cannot purely sure a checked type exception will not happen.  Let's not make that assumption. 
Besides, an originally empty caCertPEMList, and an empty caCertPEMList because of certificate convert failure are two different cases.  We need to distinguish them. Not sure if the backward comparability handling on client side can cover them both. Even if it can, it's not a recommended behavior.

Reply:
Looking at it once more I relaized, that this local optimization does not really mean a thing, as other methods might throw an exception and still fail early. So instead of returning null, the exception is re-thrown as runtimeexception.